Slim/Control/Queries.pm

```diff
@@ -493,6 +496,44 @@ sub albumsQuery {
 			push @{$p}, $year;
 		}
 
+		if (defined $fromSearch) {
+			# If we've got here from a search, we don't want to show the album unless it matches all the user's search criteria.
+			# This matters for a Works search: we've shown the user a Work because it matches their criteria, but it is possible
+			# that not all albums containing the Work match the all the criteria. (In this context, a work is a group of albums.)
+			if ( Slim::Schema->canFulltextSearch ) {
+				$fromSearch =~ s/ /* /g;
+				$fromSearch .= '*';
+				$fromSearch = "type:album " . $fromSearch;
+				push @{$w}, "EXISTS (select * FROM fulltext WHERE SUBSTR(fulltext.id, 33)=albums.id AND fulltext MATCH ?)";
+				push @{$p}, $fromSearch;
+			} else {
+				my $strings = Slim::Utils::Text::searchStringSplit($fromSearch);
+				if ( ref $strings->[0] eq 'ARRAY' ) {
+					push @{$w}, '(' . join( ' OR ', map { 'albums.titlesearch LIKE ?' } @{ $strings->[0] } ) . ')';
+					push @{$p}, @{ $strings->[0] };
+				}
+				else {
+					push @{$w}, 'albums.titlesearch LIKE ?';
+					push @{$p}, @{$strings};
+				}
+			}
+		}
```

Comment:
I'm still struggling with the `$fromSearch` concept. If we came from a search, then we'd come here because this album was part of the search's result set, right? So why would we filter by search criteria again? Wouldn't we have the album ID? And why would this be different from searching for albums or artists where we don't have this `$fromSearch`? Filtering by re-creating a full text search looks so wrong IMHO. That should never happen unless we do the actual search, not when drilling down to search results.

Reply:
I'll try to explain more fully, through an example.

Let's say the user searches for "Zinman" and has his album conducting the Zurich Tonhalle Orchestra in Beethoven's 5th and 6th Symphonies.

This will bring up results for David Zinman in Artists (obviously), the Album in Albums and individual tracks in Songs. In these cases, no further refinement is necessary.

But the new functionality will also show, under the Works subheading, the two works: Beethoven 5 and Beethoven 6. Now, say the user selects Beethoven 5. We don't want to show all versions of Beethoven's 5th Symphony they have in their library, we want to show only the one(s) where Zinman is the conductor.

This is in contrast to the situation where the search was for "Beethoven 5" - in that case the Works result for Beethoven's 5th Symphony should lead to all versions of that work that they have in their library.

So even though the user gets the same result (Beethoven's 5th) under Works, the results they need are different depending on the search criteria that they started with.

It is more complicated than Album or Song results, as each of those resolves easily to a single row of the albums table or tracks table respectively, but a work can exist in multiple albums as well as spanning multiple tracks.

I hope this makes it clear.

Reply:
But it's still similar to eg. "Filter Contributor Roles at the Album & Track Level", isn't it? Deciding whether when drilling down some entity, we want the items below that entity have filtered the same way?

TBH: I believe the no* implementations are broken in some places. But shouldn't we handle them all the same? This is a new logic trying to solve the same problem, isn't it? And I believe that the more implementations we have for the same problem, the more inconsistencies we'll introduce...

Reply:
You're right about no role filter etc. Even if they're not broken, they're very confusing. But that's not relevant here anyway.

This issue here is with the live JS search: It uses the entered search criteria, presenting results in currently 3 categories (Artists, Albums and Songs). 

Looking at the Javascript, Artist and Album links go to clixmlbrowser with the artist or album id respectively. By definition, that completely defines what the user is looking for - an artist or an album which matches their criteria. (The Song link goes directly to the songinfo.html - again unambiguous)

But with the new Works category, we cannot resolve the search with only the work id, because that work id may pull in multiple albums, with not all of them relevant to the original search criteria. So I need to pass additionally the original search term(s) to filter the detailed album list.

In an early version of this, I had an entry in the Works category for each album that matched - for example if we had multiple albums containing Beethoven 5 conducted by Zinman, the Zinman search would show multiple "Beethoven 5" results in the Works category.

Technically, that has the advantage of giving us a full key for each one to pass on to clixmlbrowser (the album id), but it is horrible from a UI point of view, especially because the live search JS pop-up is very small and width restricted - it's nowhere near as usable as the proper album menu list. 

And to properly display the album list with the correct entries, we need both the work id *and* the original search criteria in albumsQuery.

In short: the work id alone does not uniquely identify the required item(s), so we have to pass additional data.

I'm sure there would be other ways to do this, there always are, but all solutions would require the Javascript to pass on the original search criteria as well as the identified Work Id, and the current code has the big advantage of running the same query that resulted in the Work being present in the JS results in the first place, so it should always return the expected results.

Reply:
Thanks again. You've explained that to me before. I should finally understand... But! (there's always a but!):

> And to properly display the album list with the correct entries, we need both the work id and the original search criteria in albumsQuery.

What use is the original search criteria if you treat it differently than in the original search? The search you implement here has very, very little in common with the actual search, hasn't it? Wouldn't you have to go through the search query again to get the actual results? Or re-implement all of it here (library!)? 

Would there be a way to extend the search query to return not only the work IDs, but also eg. the artist ID or album ID of the results? So we'd have everything to filter here to not rely on a new, specific (and only partially implemented) copy of the search function?

Reply:
I'll have to play with this. Alas: I doubt I have any works information anywhere in my files...

Reply:
So... the search did find some work, but when I clicked it, I received an empty result. I'm trying to understand what part of the resulting SQL is causing the failure.

Reply:
so... search only searches on the works table, without checking whether there's a composer for the work. But the `works` query would only consider a work a work, if it had a composer. In my case that work doesn't have a composer. Therefore it shows up in the search, but not in the works query.

Reply:
> What use is the original search criteria if you treat it differently than in the original search? The search you implement here has very, very little in common with the actual search, hasn't it? Wouldn't you have to go through the search query again to get the actual results? Or re-implement all of it here (library!)?

Original Search asks "What works do we have where the full text search = David Zinman?"
Albums Query asks "For this given work, which albums containing the work has full text search = David Zinman?"

That's why I pass "David Zinman" as well as the works id to albumsQuery.

I'm struggling to see the hole in this?

> Would there be a way to extend the search query to return not only the work IDs, but also eg. the artist ID or album ID of the results? So we'd have everything to filter here to not rely on a new, specific (and only partially implemented) copy of the search function?

I suppose another way to do it would be for the query invoked by the Javascript to return the work Id and an array of album Ids matching the original query and pass those to albumsQuery. But I'm not sure what we would gain by this.

Reply:
> so... search only searches on the works table, without checking whether there's a composer for the work. But the `works` query would only consider a work a work, if it had a composer. In my case that work doesn't have a composer. Therefore it shows up in the search, but not in the works query.

By definition, works have composers. Did you tag some files and import, or hack the database? If from tag imports, there's a hole there - I should set the composer to "unknown" or something if a track has a work with no composer.

Reply:
There are some test files I attached in the very first message in this discussion, which might help.

Reply:
It's from tags. I actually deleted the library.db because of the migration issues and started over. So there even wasn't an outdated works related schema. Really just poor tagging to break things 😁.

Should the scanner check for the existence of a composer before creating the work? Or do a cleanup call at the end (don't we do something similar for albums?).

But maybe we should just leave this as it is for now, get things merged, and concentrate on those fixes after this hits a broader audience? There will be more issues anyway, and they could be addressed in individual bug reports, instead of this huge PR (soon reaching 200 comments...).

Reply:
> Original Search asks "What works do we have where the full text search = David Zinman?"
Albums Query asks "For this given work, which albums containing the work has full text search = David Zinman?"
>
> That's why I pass "David Zinman" as well as the works id to albumsQuery.
>
> I'm struggling to see the hole in this?

The hole _I_ am seeing is that the original search applied a different logic than you apply here. It would eg. respect libraries. Just saying that re-implementing something which should result in the same outcome, should not be done. Re-use an existing implementation, or it's asking for trouble. I'm tempted to just merge this PR to be able to move forward. But I'm convinced that sooner or later this will cause problems/reports. It feels wrong to me. 

What do you prefer? Merge and look back, or address before the merge?

Reply:
Maybe I didn't make it clear that the `$fromsearch` parameter triggers an **addition** to, not replacement of, the selection criteria in `albumsQuery`, so the library check and everything else is still there:

[24-04-05 15:05:22.8176] Slim::Control::Queries::albumsQuery (717) Albums query: SELECT albums.artwork AS 'albums.artwork', albums.contributor AS 'albums.contributor', albums.disc AS 'albums.disc', albums.discc AS 'albums.discc', albums.extid AS 'albums.extid', albums.id AS 'albums.id', albums.title AS 'albums.title', albums.titlesearch AS 'albums.titlesearch', albums.titlesort AS 'albums.titlesort', albums.year AS 'albums.year', composer.name AS 'composer.name', contributors.name AS 'contributors.name', tracks.grouping AS 'tracks.grouping', tracks.work AS 'tracks.work', works.title AS 'works.title' FROM albums JOIN tracks ON tracks.album = albums.id JOIN works ON tracks.work = works.id JOIN contributors AS composer ON works.composer = composer.id JOIN contributor_track ON contributor_track.track = tracks.id JOIN contributors ON contributors.id = albums.contributor WHERE albums.id IN (SELECT library_album.album FROM library_album WHERE library_album.library = ?) **AND EXISTS (select * FROM fulltext WHERE SUBSTR(fulltext.id, 33)=albums.id AND fulltext MATCH ?)** AND tracks.work = ? AND contributor_track.contributor = ? AND contributor_track.role = 2 GROUP BY tracks.grouping, albums.id ORDER BY albums.titlesort COLLATE en_US , albums.disc LIMIT ?,?  / ["91b427f1", "**type:album zinman***", 13, 21, 0, 50]

Sorry about that, but in any case, yes, we should merge now, if only to make the discussion more manageable, I'm not convinced I've responded to all your points, or even seen them all, in this massive comments section (I can't see a way to view it in simple chronological sequence).

I've just pushed the latest changes. let me if I've missed anything essential.